api/v1alpha1/extension_types.go

```diff
@@ -60,11 +64,17 @@ type ExtensionSourcePackage struct {
 	UpgradeConstraintPolicy UpgradeConstraintPolicy `json:"upgradeConstraintPolicy,omitempty"`
 }
 
+// +kubebuilder:validation:XValidation:rule="self.sourceType=='package' && has(self.__package__)",message="sourceType must match populated union field"
+//
 // ExtensionSource defines the source for this Extension, right now, only a package is supported.
 type ExtensionSource struct {
-	// A source package defined by a name, version and/or channel
+	//+kubebuilder:validation:Enum:=package
 	//+kubebuilder:validation:Required
-	Package *ExtensionSourcePackage `json:"package"`
+	// sourceType is the discriminator for the source type
+	SourceType string `json:"sourceType"`
+
+	// package defines a reference for a bundle in a catalog defined by a name and a version and/or channel
+	Package *ExtensionSourcePackage `json:"package,omitempty"`
```

Comment:
Maybe not the right place to comment this, but I think we should add a kustomize patch to enforce a required `oneOf` with `package` being the only entry for now. This seems within scope of this PR to still enforce that more than the `sourceType` field is required.

Similar to: https://github.com/operator-framework/catalogd/blob/main/config/crd/patches/catalog_validation.yaml

Reply:
Oh, I see now the CEL validation handles that - disregard :)